bootstrap/kubeadm/internal/ignition/ignition.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package ignition aggregates all Ignition flavors into a single package to be consumed
+// by the bootstrap provider by exposing an API similar to 'internal/cloudinit' package.
+package ignition
+
+import (
+	"fmt"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/cloudinit"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/internal/ignition/clc"
+)
+
+const (
+	joinSubcommand         = "join"
+	initSubcommand         = "init"
+	kubeadmCommandTemplate = "kubeadm %s --config /etc/kubeadm.yml %s"
```

Comment:
please make sure the ignition format also follows all requirements in the bootstrap provider contract, in particular, writing the sentinel file: https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/bootstrap.md#sentinel-file

Reply:
Would `mkdir -p /run/cluster-api && touch /run/cluster-api/bootstrap-success.complete` after `kubeadm init` be sufficient?

Reply:
It should. In cloud init we do `kubeadm init/join && echo success > /run/cluster-api/bootstrap-success.complete` but that's because it has to be compatible with windows https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/internal/cloudinit/cloudinit.go#L32

Reply:
Added. I'm not sure what's the best way to test this though.

Reply:
for CABPK it's tested with CAPD in e2e tests. How's the ignition stuff going to be tested? Are we going to build any ignition clusters in e2e PR tests?

Reply:
>How's the ignition stuff going to be tested? Are we going to build any ignition clusters in e2e PR tests?

I'm not familiar with CAPI e2e test framework, so some suggestions would be very welcome here :) The testing we made so far was done using CAPA e2e test framework.

Reply:
The CAPI e2e tests use the Doker infrastructure provider (CAPD). I would recommend adding an e2e test for ignition to prevent regressions. In terms of PR gates, the test shouldn't run on every PR but ideally it would be triggered for every PR that touches this part of the code.